src/syntax/sedlex.ml

```diff
@@ -117,6 +129,89 @@ let transition (state : state) =
   Array.sort (fun (c1, _) (c2, _) -> compare c1 c2) t;
   t
 
+(* Restore NFA path from DFA path *)
+
+type trans_case = {
+  curr_state : int;
+  curr_node : int;
+  prev_state : int;
+  prev_node : int;
+  char_set : Sedlex_cset.t;
+  actions : action list;
+}
+
+type final_case = { curr_node : int; actions : action list }
+
+let compile_traces states (start, final) =
+  let counter = ref 0 in
+  let nodes_idx = Hashtbl.create 31 in
+  let rec aux node =
+    if not (Hashtbl.mem nodes_idx node) then begin
+      let i = !counter in
+      incr counter;
+      Hashtbl.add nodes_idx node i;
+      List.iter aux node.eps;
+      List.iter (fun (_, next) -> aux next) node.trans
+    end
+  in
+  aux start;
+  let append_action actions = function
+    | None -> actions
+    | Some action -> action :: actions
+  in
+  let first_node = Hashtbl.find nodes_idx final in
+  let trans_cases =
+    let cases = Hashtbl.create 31 in
+    Hashtbl.iter
+      (fun from_state j ->
+        Hashtbl.iter
+          (fun to_state i ->
+            List.iter
+              (fun from_node ->
+                try
+                  let node_j = Hashtbl.find nodes_idx from_node in
+                  let rec dfs cset actions to_node =
+                    let node_i = Hashtbl.find nodes_idx to_node in
+                    if not (Hashtbl.mem cases (i, node_i, j, cset)) then begin
+                      let actions = append_action actions to_node.action in
+                      if to_node.trans <> [] || to_node == final then
+                        Hashtbl.add cases (i, node_i, j, cset)
+                          {
+                            curr_state = i;
+                            curr_node = node_i;
+                            prev_state = j;
+                            prev_node = node_j;
+                            char_set = cset;
+                            actions;
+                          };
+                      List.iter (dfs cset actions) to_node.eps
+                    end
+                  in
+                  List.iter
+                    (fun (cset, to_node) ->
+                      if List.mem to_node to_state then dfs cset [] to_node)
+                    from_node.trans
+                with Not_found -> ())
+              from_state)
+          states)
+      states;
+    Hashtbl.to_seq_values cases |> List.of_seq
+  in
+  let final_cases =
+    let rec dfs actions cases node =
+      let i = Hashtbl.find nodes_idx node in
+      let actions = append_action actions node.action in
+      let cases =
+        if node.trans <> [] || node == final then
```

Comment:
Can you add a comment explaining these conditions ?

Reply:
Will add it later.